### PR TITLE
Implementing rate limit

### PR DESCRIPTION
**Pull request by mockthebear**

### CODE REVIEW

File: docs/resources/gocache_smart_rules_ratelimit.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
+# gocache_smart_rules_ratelimit
+
+Provides a list of type `ratelimit` Smart Rules for a domain, which effectively are just rate limit rules. Smart Rules are the GoCache platform for transforming the request behavior under selected conditions and the type `ratelimit` are like the other smart rules with the very same matching criteria just changing the actions to define rate limiting actions. See more https://docs.gocache.com.br/ratelimit/.
```
Comment (morguesa) [inline docs/resources/gocache_smart_rules_ratelimit.md]: behavio**u**r

and the type `ratelimit` **is** like

File: docs/resources/gocache_smart_rules_ratelimit.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+```
+
+## Argument Reference
+
+The `gocache_smart_rules_ratelimit` argument layout is a complex structure composed of several sub-resources - these resources are laid out below.
+
+### Top-Level Arguments
+
+* `domain` - (Required) The domain where you will apply the policies.
+* `smart_rule` - (Required) Each `smart_rule` block sets a rule. More than one block is allowed. The list have no particular order in contrast to other smartrules, each of those rules defined in the block are evaluated.
```
Comment (morguesa) [inline docs/resources/gocache_smart_rules_ratelimit.md]: The list **has** ...

File: docs/resources/gocache_smart_rules_ratelimit.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+* `match` - (Required) Request criteria needed to trigger the rule action. At least one option is needed. See the options in [match level arguments](match-level-arguments).
+* `action` - (Required) Actions applied to a request when the criteria is matched. At least one option is needed. See the options in [action level arguments](action-level-arguments).
+* `metadata` - (Required) Rule metadata. See the options in [metadata level arguments](metadata-level-arguments).
+
+### Match Level Arguments
+
+Some fields can use regex patterns. To know more about regex, see the documentation in https://docs.gocache.com.br/smart_rules/smart_rules-regexp/.
+
+* `request_method` - (Optional) List of HTTP request methods matched by the rule. Possible values `GET`, `POST`, `PUT`, `DELETE`, `HEAD`, `OPTIONS`.
```
Comment (morguesa) [inline docs/resources/gocache_smart_rules_ratelimit.md:75]: PATCH is not a option?

File: docs/resources/gocache_smart_rules_ratelimit.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+* `scheme` - (Optional) List of HTTP request scheme matched by the rule. Possible values `http`, `https` or `http*`
+* `hostname` - (Optional) HTTP request hostname matched by the rule. For all hostnames, set `*`.
+* `request_uri` - (Optional) HTTP request URI pattern matched by the rule. You can use regex in this field.
+* `http_user_agent` - (Optional) Request user agent matched by the rule. You can use regex in this field.
+* `cookie` - (Optional) List of cequest cookie names matched by the rule. This field supports wildcard.
```
Comment (morguesa) [inline docs/resources/gocache_smart_rules_ratelimit.md]: **r**equest cookie

File: docs/resources/gocache_smart_rules_ratelimit.md [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+* `cookie_content` - (Optional) Cookie values for a cookie matched by the rule. You need to declare cookie and its value separeted by `=`, like `wp_test=yes`. You can use regex in the value.
```
Comment (morguesa) [inline docs/resources/gocache_smart_rules_ratelimit.md]: separ**a**ted

File: docs/resources/gocache_smart_rules_ratelimit.md [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+* `http_version` - (Optional) List of request HTTP version matched by the rule. Possible values are `HTTP/1.0`,`HTTP/1.1`,`HTTP/2.0`.
+* `header` - (Optional) HTTP headers matched by the rule. Header name and value are separated by `:`. You can use regex in the value.
+* `origin_country` - (Optional) List of origin countries matched by the rule in the ISO format, like `BR`.
+* `origin_continent` - (Optional) List of origin continents matched by the rule in the ISO format. Possible values `SA`, `NA`, `OC`, `EU`, `AF`, `AS`.
+* `device_type` - (Optional) list of device types based on user agent matched by the rule. Possible values `mobile`,`desktop`,`bot`,`na`
+* `bots` - (Optional) If the bot is legitimate. Possible values `known`, `others`
+
+### Action Level Arguments
+
+* `rate_limit_action` - (Required) When `block`, the matched request are denied with status `429`, when `challenge`, a captcha challenge is returned and when `simulate`, the request is allowed, even when WAF blocks it. Possible values: `block`, `challenge`, `simulate`
```
Comment (morguesa) [inline docs/resources/gocache_smart_rules_ratelimit.md]: the matched request **is** denied

File: docs/resources/gocache_smart_rules_ratelimit.md [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+* `rate_limit_amount` - (Required) Amount of requests from a single ip needed to the action to be taken.
```
Comment (morguesa) [inline docs/resources/gocache_smart_rules_ratelimit.md]: a single ip needed **for** the action to be taken.

File: docs/resources/gocache_smart_rules_ratelimit.md [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+* `rate_limit_period` - (Required) Defines the minimum time windows (seconds) to the requests reach `rate_limit_amount` so the action can be taken. Possible values `86400`, `43200`, `14400`, `3600`, `1800`, `600`, `300`, `120`, `60`, `30`, `10`
```
Comment (morguesa) [inline docs/resources/gocache_smart_rules_ratelimit.md]: Defines the minimum time **window** (seconds) for the requests reach `rate_limit_amount` **for** the action can be taken

File: gocache/singledomain.go [hunk 9]
```diff
@@ -0,0 +1,118 @@
+package gocache
+
+// import (
```
Comment (morguesa) [inline gocache/singledomain.go]: Remove this file

**Review by morguesa — CHANGES_REQUESTED** [review]

File: docs/resources/gocache_smart_rules_ratelimit.md (hunk 3 above)
Comment (mockthebear) [reply docs/resources/gocache_smart_rules_ratelimit.md:75]: no, its not

**Review by morguesa — DISMISSED** [review]

File: docs/resources/gocache_smart_rules_ratelimit.md [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+### Smart Rule Level Arguments
+
+* `match` - (Required) Request criteria needed to trigger the rule action. At least one option is needed. See the options in [match level arguments](match-level-arguments).
+* `action` - (Required) Actions applied to a request when the criteria is matched. At least one option is needed. See the options in [action level arguments](action-level-arguments).
+* `metadata` - (Required) Rule metadata. See the options in [metadata level arguments](metadata-level-arguments).
+
+### Match Level Arguments
+
+Some fields can use regex patterns. To know more about regex, see the documentation in https://docs.gocache.com.br/smart_rules/smart_rules-regexp/.
```
Comment (Mewbi) [inline docs/resources/gocache_smart_rules_ratelimit.md]: Instead raw hyperlink.

```
...To know more about regex, check the documentation about [Smart Rules Regexp](https://docs.gocache.com.br/smart_rules/smart_rules-regexp/)
```

File: docs/resources/gocache_smart_rules_ratelimit.md [hunk 11]
```diff
@@ -0,0 +1,109 @@
+# gocache_smart_rules_ratelimit
+
+Provides a list of type `ratelimit` Smart Rules for a domain, which effectively are just rate limit rules. Smart Rules are the GoCache platform for transforming the request behaviour under selected conditions and the type `ratelimit` are like the other smart rules with the very same matching criteria just changing the actions to define rate limiting actions. See more https://docs.gocache.com.br/ratelimit/.
```
Comment (Mewbi) [inline docs/resources/gocache_smart_rules_ratelimit.md]:
```
... See more in [Rate Limit documentation](https://docs.gocache.com.br/ratelimit/).
```

**Review by Mewbi — CHANGES_REQUESTED** [review]

**Review by morguesa — APPROVED** [review]

**Review by Termack — APPROVED** [review]

**Review by Mewbi — APPROVED** [review]

**Review by Mewbi — APPROVED** [review]

**Review by Mewbi — APPROVED** [review]

**Review by Mewbi — APPROVED** [review]

**Pull request merged**